### PR TITLE
Segment out meta ops from a fusion. 

### LINKED ISSUE #2395: RuntimeError: producer->getMemoryType() == MemoryType::Global
This blocks https://github.com/Lightning-AI/lightning-thunder/issues/191. 

# nvFuser reproducer

Check out https://github.com/NVIDIA/Fuser/tree/wjy/memory and run `python repro.py`. 

```py
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id9(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, None], dtype=DataType.Bool, is_cpu=False, stride_order=[2, 1, 0])
    T1 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.Float, is_cpu=False, stride_order=[2, 1, 0])
    T2 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Bool, is_cpu=False, stride_order=[1, 0])
    S3 = fd.define_scalar(16, dtype=DataType.Int)
    S4 = fd.define_scalar(16, dtype=DataType.Int)
    S5 = fd.define_scalar(1, dtype=DataType.Int)
    V6 = fd.define_vector([S3, S4, S5], dtype=DataType.Int)
    T7 = fd.ops.broadcast_in_dim(T2, shape=V6, broadcast_dims=[0, 1])
    S8 = fd.define_scalar(1, dtype=DataType.Int)
    S9 = fd.define_scalar(16, dtype=DataType.Int)
    S10 = fd.define_scalar(1, dtype=DataType.Int)
    S11 = fd.define_scalar(16, dtype=DataType.Int)
    S12 = fd.define_scalar(32, dtype=DataType.Int)
    S13 = fd.define_scalar(1, dtype=DataType.Int)
    V14 = fd.define_vector([S8, S9, S10, S11, S12, S13], dtype=DataType.Int)
    T15 = fd.ops.broadcast_in_dim(T7, shape=V14, broadcast_dims=[1, 3, 5])
    S16 = fd.define_scalar(16, dtype=DataType.Int)
    S17 = fd.define_scalar(16, dtype=DataType.Int)
    S18 = fd.define_scalar(32, dtype=DataType.Int)
    V19 = fd.define_vector([S16, S17, S18], dtype=DataType.Int)
    T20 = fd.ops.reshape(T15, new_shape=V19)
    T21 = fd.ops.slice(T1, start_indices=[0, 0, 16], end_indices=[16, 16, 32], strides=[1, 1, 1])
    S22 = fd.define_scalar(0.00000, dtype=DataType.Double)
    T23 = fd.ops.where(T20, S22, T1)
    S24 = fd.define_scalar(0.00000, dtype=DataType.Double)
    T25 = fd.ops.where(T0, S24, T23)
    fd.add_output(T21)
    fd.add_output(T25)

with FusionDefinition() as fd:
    nvfuser_fusion_id9(fd)

inputs = [
    torch.randint(0, 2, (256,), dtype=torch.bool, device='cuda:1').as_strided((16, 16, 32), (16, 1, 0)),
    torch.randn((8192,), dtype=torch.float32, device='cuda:1').as_strided((16, 16, 32), (512, 32, 1)),
    torch.randint(0, 2, (256,), dtype=torch.bool, device='cuda:1').as_strided((16, 16), (16, 1)),
]
fd.execute(inputs)
```

```
Traceback (most recent call last):
  File "/opt/pytorch/nvfuser/nvfuser/__init__.py", line 145, in execute
    result = self._execute(
RuntimeError: producer->getMemoryType() == MemoryType::Global INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/device_lower/analysis/sync_information.cpp":699, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Inconsistent parallelization found between TV12 (T12_l[ iblockIdx.x180{( ceilDiv(( ceilDiv(( ceilDiv(( 16 * 16 ), 128) ), 1) ), 1) )}, iUS181{1}, iS179{1}, ithreadIdx.x177{128} ]) and TV3(T3_l[ iblockIdx.x173{( ceilDiv(( ceilDiv(( ceilDiv(( 16 * 16 ), 128) ), 1) ), 1) )}, iUS174{1}, iS172{1}, ithreadIdx.x170{128}, bS10{1} ] ca_pos( 4 )). Producer is required to be in Global Memory based on parallelization strategy. RAW flags: (blockIdx.x threadIdx.x)
Exception raised from SyncMap at /opt/pytorch/nvfuser/csrc/device_lower/analysis/sync_information.cpp:699 (most recent call first):
frame #0: nvfuser::nvfCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7742424c9149 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #1: nvfuser::nvfErrorFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x53 (0x7742427fc8a3 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #2: nvfuser::SyncMap::SyncMap(nvfuser::Fusion*) + 0x24a4 (0x77424270e0c4 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #3: <unknown function> + 0x37f560 (0x77424271e560 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #4: nvfuser::GpuLower::GpuLower(nvfuser::Fusion*, nvfuser::CompileParams const&) + 0xca8 (0x77424271fb48 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #5: nvfuser::FusionExecutor::compileFusion(nvfuser::Fusion*, nvfuser::KernelArgumentHolder const&, nvfuser::LaunchParams const&, nvfuser::CompileParams, nvfuser::ScheduleHeuristic, long, long, long, long) + 0x3ed (0x77424281417d in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #6: <unknown function> + 0x65dd42 (0x7742429fcd42 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #7: nvfuser::FusionKernelRuntime::compileFusionParallel(nvfuser::KernelArgumentHolder) + 0x492 (0x774242a05fb2 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #8: nvfuser::FusionExecutorCache::runFusionWithInputs(c10::ArrayRef<c10::IValue> const&, std::optional<nvfuser::PrimDataType>, std::optional<signed char>) + 0xad3 (0x774242a10eb3 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #9: nvfuser::python_frontend::FusionDefinition::execute(c10::ArrayRef<c10::IValue> const&, std::optional<signed char>, bool, bool, bool) const + 0x3ec (0x774242c0264c in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #10: <unknown function> + 0x1a42ce (0x7742425432ce in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #11: <unknown function> + 0x21ae3f (0x7742425b9e3f in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
frame #12: <unknown function> + 0x2aeb60 (0x77424264db60 in /opt/pytorch/nvfuser/nvfuser/_C.cpython-310-x86_64-linux-gnu.so)
<omitting python frames>
frame #28: <unknown function> + 0x29d90 (0x774351ea1d90 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #29: __libc_start_main + 0x80 (0x774351ea1e40 in /usr/lib/x86_64-linux-gnu/libc.so.6)
```

# Thunder reproducer

Check out https://github.com/Lightning-AI/lightning-thunder/tree/wjy/bookend and run `python thunder/tests/distributed/test_tensor_parallel.py -k TensorParallelTest.test_both_column_and_row_bias_True`

### PR DESCRIPTION
**Pull request by wujingyue**
We currently do that to some extent by adding `segment_set` before aliased outputs. This PR modifies and extends the logic for aliased inputs as well. 

Fixes #2395

### CODE REVIEW

**Comment by wujingyue** [comment]
!build

**Comment by wujingyue** [comment]
!build

File: csrc/fusion.cpp [hunk 1]
```diff
@@ -320,7 +320,11 @@ void Fusion::replaceOutput(Val* output, Val* replacement) {
     }
     if (output->getValType().value() == ValType::TensorView) {
       output->setIsFusionOutput(false);
-      output->as<TensorView>()->setMemoryType(MemoryType::Local);
+      // If `output` is both an input and an output before the replacement,
+      // don't localize it.
+      if (!output->isFusionInput()) {
+        output->as<TensorView>()->setMemoryType(MemoryType::Local);
```
Comment (jjsjann123) [inline csrc/fusion.cpp:326]: :+1:

File: csrc/preseg_passes/mark_aliases_prepare.cpp [hunk 2]
```diff
@@ -49,62 +51,82 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
...
-    //
-    //   N/M -> M1 -> M2
-    //      |
-    //      --> M1' (non-output copy of M1) -> N/M
-    //
-    // and then put a `segment_set` on N/M->M1.
-    aliased_out->cacheBefore(LoadStoreOpType::SegmenterSet);
+    // Some users of `aliased_io` are used by non-aliases, and some are used
+    // only by aliases. We put a `segment_set` after `aliased_io` and redirect
+    // the latter group to use the `segment_set`.
```
Comment (jjsjann123) [inline csrc/preseg_passes/mark_aliases_prepare.cpp:122]: qq: if aliased_io is a fusion input, shouldn't we skip this as well?

File: csrc/preseg_passes/mark_aliases_prepare.cpp [hunk 3]
```diff
@@ -49,62 +51,82 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
...
+
+    // If all users are used by non-alias outputs, do nothing. Adding
+    // segment_set around unlikely creates more no-op regions.
+    if (users_used_only_by_aliases.empty()) {
+      continue;
+    }
+
+    // If all users are used by aliases, put a `segment_set` before it.
+    if (users_used_by_non_aliases.empty()) {
+      if (aliased_io->isFusionInput()) {
```
Comment (jjsjann123) [inline csrc/preseg_passes/mark_aliases_prepare.cpp:104]: I think this one short-cut should be lifted.

File: tests/cpp/test_alias.cpp [hunk 4]
```diff
@@ -1325,4 +1355,44 @@ TEST_F(AliasTest, FusionExecutor) {
...
+  fusion->addOutput(permute_out);
+  fusion->addOutput(compute_out);
+
+  FusionExecutorCache fec(std::move(fusion));
+  at::Tensor in_tensor = at::randn({2, 3}).cuda();
+  std::vector<at::Tensor> out_tensors = fec.runFusionWithInputs({in_tensor});
+  testValidate(fec.fusion(), out_tensors, {in_tensor}, __LINE__, __FILE__);
+
+  at::Tensor slice_out_tensor = out_tensors[0];
+  EXPECT_TRUE(slice_out_tensor.is_alias_of(in_tensor));
```
Comment (jjsjann123) [inline tests/cpp/test_alias.cpp:1376]: nitpick:
```suggestion
  at::Tensor permute_out_tensor = out_tensors[0];
  EXPECT_TRUE(permute_out_tensor.is_alias_of(in_tensor));
```

File: tests/cpp/test_preseg_passes.cpp [hunk 5]
```diff
@@ -631,4 +624,22 @@ TEST_F(NVFuserTest, FusionRemoveEmptyMatmul_CUDA) {
   testValidate(preseg_fusion, outputs, aten_inputs, __LINE__, __FILE__);
 }
 
+TEST_F(PresegTest, ReplaceOutput) {
```
Comment (jjsjann123) [inline tests/cpp/test_preseg_passes.cpp:627]: :clap:

**Review by jjsjann123 — APPROVED** [review]
LGTM with minor comments

File: csrc/preseg_passes/mark_aliases_prepare.cpp (hunk 2 above)
Comment (wujingyue) [reply csrc/preseg_passes/mark_aliases_prepare.cpp:122]: I believe this is still needed for fusion inputs and AliasTest.SegmentMetaOps is an example. There, a fusion input is used by a permute on one branch and a series of compute op on the other branch. The code here adds a segment_set between the fusion input and the permute, separating the permute and the compute.

File: csrc/preseg_passes/mark_aliases_prepare.cpp (hunk 3 above)
Comment (wujingyue) [reply csrc/preseg_passes/mark_aliases_prepare.cpp:104]: I think https://github.com/NVIDIA/Fuser/pull/2529/files#r1669416615 answers this as well.

**Comment by wujingyue** [comment]
!build

File: csrc/preseg_passes/mark_aliases_prepare.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/preseg_passes/mark_aliases_prepare.cpp:122]: I see... The segmenter set is necessary for the segmenter logic.

**Pull request merged**